tests/test/opaque_types.rs

```diff
@@ -0,0 +1,105 @@
+use super::*;
+
+#[test]
+fn opaque_bounds() {
+    test! {
+        program {
+            struct Ty { }
+
+            trait Clone { }
+            opaque type T: Clone = Ty;
+        }
+
+        goal {
+            T: Clone
+        } yields {
+            "Unique; substitution []"
+        }
+    }
+}
+
+#[test]
+fn opaque_reveal() {
+    test! {
+        program {
+            struct Ty { }
+            trait Trait { }
+            impl Trait for Ty { }
+
+            trait Clone { }
+            opaque type T: Clone = Ty;
+        }
+
+        goal {
+            if (Reveal) {
+                T: Trait
+            }
+        } yields {
+            "Unique; substitution []"
+        }
+
+        goal {
+            T: Trait
+        } yields {
+            "No possible solution"
+        }
+    }
+}
+
+#[test]
+fn opaque_generics_simple() {
+    test! {
+        program {
+            trait Iterator { type Item; }
+
+            struct Vec<T> { }
+            struct Bar { }
+            impl<T> Iterator for Vec<T> {
+                type Item = u32;
+            }
+
+            opaque type Foo<X>: Iterator = Vec<X>;
+        }
+
+        goal {
+            Foo<Bar>: Iterator
+        } yields {
+            "Unique; substitution []"
+        }
+
+    }
+}
+
+#[test]
+fn opaque_generics() {
+    test! {
+        program {
+            trait Iterator { type Item; }
+
+            struct Vec<T> { }
+            struct Bar { }
+            impl<T> Iterator for Vec<T> {
+                type Item = T;
+            }
+
+            opaque type Foo<X>: Iterator<Item = X> = Vec<X>;
+        }
+
+        goal {
+            Foo<Bar>: Iterator<Item = Bar>
+        } yields {
+            "Unique; substitution []"
+        }
+
+        goal {
+            forall<T> {
+                if (T: Iterator) {
```

Comment:
why the `T: Iterator` bound here?

Reply:
Right, this bound can be assumed, because we're assuming here that `Foo<X>` is well-formed.
Actually checking that `T` implements `Iterator` should only be done when you actually use `Foo`, at that point the WF-check (which isn't implemented yet for impl trait in chalk) checks the `T: Iterator` bound.

Reply:
Well, not even that. `T` is the type parameter, there's no `Iterator` bound or reason why it would need to be an `Iterator` anywhere (`Foo<T>` is the `Iterator`).

Reply:
Whoops you're right, that bound is totally unnecessary because `T` is the `Item`, not the `Iterator`.